src/sort_by_hash.h

```diff
@@ -0,0 +1,51 @@
+#include <vector>
+#include "datatypes.h"
+#include "hash_integers.h"
+
+/*
+  Sort an array of particles by destination rank, where the destination MPI
+  rank is determined by (HashInteger(p.Id) % comm_size). The input Particle_t type
+  should have an Id field.
+
+  Returns an array with the offset to the first particle to go to each rank.
+*/
+template <typename Particle_t>
+std::vector<HBTInt> sort_by_hash(std::vector<Particle_t> &Particles, int comm_size)
+{
+
+  // First we need to count the number of elements to send to each rank
+  std::vector<HBTInt> count(comm_size, 0);
+  for (HBTInt i = 0; i < Particles.size(); i += 1)
+  {
+    int dest = RankFromIdHash(Particles[i].Id, comm_size);
+    count[dest] += 1;
+  }
+
+  // Compute offset to the first particle for each rank in the output
+  std::vector<HBTInt> offset(comm_size, 0);
+  for (int i = 1; i < comm_size; i += 1)
+  {
+    offset[i] = offset[i - 1] + count[i - 1];
+  }
+
+  // Allocate storage for the sorted array
+  std::vector<Particle_t> SortedParticles(Particles.size());
+
+  // Reset the counts
+  std::fill(count.begin(), count.end(), 0);
+
+  // Populate the output array in sorted order
+  // TODO: can we parallelize this somehow?
```

Comment:
Not that aware of how to do OMP parallel sorts, but the following slides (7th one in particular) have some information that might be useful: https://cw.fel.cvut.cz/old/_media/courses/b4m35pag/lab6_slides_advanced_openmp.pdf

We could create a separate vector to hold RankFromIdHash (done in OMP), and then apply the above OMP parallel sorting algorithm (e.g. within your argsort function or a variant) to that vector. We then rearrange the Particle vectors. Is there a complication in doing it like this?

Reply:
I was leaning towards leaving this serial because we'd be adding complexity and a chance of bugs in return for a tiny performance improvement. It would be good to have a multi threaded sort available, but maybe not in this pull request?

Reply:
I agree, we should leave this for a potential future addition.

Reply:
Also, this function is a counting sort which runs in O(N) time so replacing it with a general purpose sort with O(N log N) running time but using threads might not always be faster.